src/tests/test_hiv_testing.py

```diff
@@ -0,0 +1,190 @@
+    # fixing some values
+    pop.hiv_testing.date_start_testing = 2003.5
+    pop.hiv_testing.eff_max_freq_testing = 1
+    pop.hiv_testing.date_test_rate_plateau = 2015.5
+    pop.hiv_testing.an_lin_incr_test = 0.8
+    pop.hiv_testing.no_test_if_np0 = False
+    pop.hiv_testing.covid_disrup_affected = False
+    pop.hiv_testing.testing_disrup_covid = False
+
+    # have roughly 20% of the population be dead
+    r = rng.uniform(size=len(pop.data))
+    dead = r < 0.2
+    pop.set_present_variable(col.DATE_OF_DEATH, pop.date, sub_pop=pop.apply_bool_mask(dead))
+
+    # evolve population
+    pop.hiv_testing.update_hiv_testing(pop)
+
+    # get stats
+    testing_population = pop.get_sub_pop([(col.HARD_REACH, op.eq, False), (col.DATE_OF_DEATH, op.eq, None)])
+    # all previously tested
+    no_repeat_testers = len(pop.get_sub_pop([(col.LAST_TEST_DATE, op.eq, pop.date)]))
+    prob_test = pop.hiv_testing.calc_prob_test(True, 0, 0)
+    mean = len(testing_population) * prob_test
+    stdev = sqrt(mean * (1 - prob_test))
+    # check tested value is within 3 standard deviations
+    assert mean - 3 * stdev <= no_repeat_testers <= mean + 3 * stdev
+    # check that no dead people were just tested
+    assert (pop.get_variable(col.DATE_OF_DEATH, sub_pop=testing_population).isna()).all()
+
+
+def test_partner_reset_after_test():
+
+    for ever_tested in [True, False]:
+
+        # build population
+        N = 100000
+        pop = Population(size=N, start_date=date(2010, 1, 1))
+        pop.data[col.AGE] = 20
+        pop.data[col.EVER_TESTED] = ever_tested
+        if ever_tested:
+            pop.data[col.LAST_TEST_DATE] = date(2008, 1, 1)
+        pop.data[col.NP_LAST_TEST] = 2
+        pop.data[col.NSTP_LAST_TEST] = 1
+        # fixing some values
+        pop.hiv_testing.date_start_testing = 2003.5
+        pop.hiv_testing.eff_max_freq_testing = 1
+        pop.hiv_testing.init_rate_first_test = 0.1
+        pop.hiv_testing.date_test_rate_plateau = 2015.5
+        pop.hiv_testing.an_lin_incr_test = 0.8
+        pop.hiv_testing.no_test_if_np0 = False
+        pop.hiv_testing.covid_disrup_affected = False
+        pop.hiv_testing.testing_disrup_covid = False
+
+        # diagnose roughly 20% of the population with HIV
+        r = rng.uniform(size=len(pop.data))
+        diagnosed = r < 0.2
+        pop.data[col.HIV_STATUS] = diagnosed
+
+        # evolve population
+        pop.hiv_testing.update_hiv_testing(pop)
+
+        # get people that were just tested
+        tested_population = pop.get_sub_pop([(col.LAST_TEST_DATE, op.eq, pop.date)])
+        # check that partner numbers have been reset
+        assert sum(pop.get_variable(col.NP_LAST_TEST, sub_pop=tested_population)) == 0
+        assert sum(pop.get_variable(col.NSTP_LAST_TEST, sub_pop=tested_population)) == 0
+        # check that no people just tested were already diagnosed with HIV
+        assert (~pop.get_variable(col.HIV_STATUS, sub_pop=tested_population)).all()
```

Comment:
This line feels like it would be more at home in a different test e.g. the first time & repeat testers checks, or in its own; I think `test_partner_reset_after_test` suggests a very specific test so if it fails on this line people might be confused about what actually is going wrong.